# Ex3 - Getting and Knowing your Data

### Step 1. Import the necessary libraries

In [ ]:
import pandas as pd

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.user).

In [ ]:
from urllib import request

URL = "https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.user"

response = request.urlretrieve(URL, "u.user")

### Step 3. Assign it to a variable called users and use the 'user_id' as index

In [ ]:
users = pd.read_csv("/content/u.user", sep = "|" , index_col = "user_id")
users

age gender     occupation zip_code
user_id                                    
1         24      M     technician    85711
2         53      F          other    94043
3         23      M         writer    32067
4         24      M     technician    43537
5         33      F          other    15213
...      ...    ...            ...      ...
939       26      F        student    33319
940       32      M  administrator    02215
941       20      M        student    97229
942       48      F      librarian    78209
943       22      M        student    77841

[943 rows x 4 columns]

### Step 4. See the first 25 entries

In [ ]:
users.head(25)

age gender     occupation zip_code
user_id                                    
1         24      M     technician    85711
2         53      F          other    94043
3         23      M         writer    32067
4         24      M     technician    43537
5         33      F          other    15213
6         42      M      executive    98101
7         57      M  administrator    91344
8         36      M  administrator    05201
9         29      M        student    01002
10        53      M         lawyer    90703
11        39      F          other    30329
12        28      F          other    06405
13        47      M       educator    29206
14        45      M      scientist    55106
15        49      F       educator    97301
16        21      M  entertainment    10309
17        30      M     programmer    06355
18        35      F          other    37212
19        40      M      librarian    02138
20        42      F      homemaker    95660
21        26      M         writer    30068
22        25      M         writer    40206
23        30      F         artist    48197
24        21      F         artist    94533
25        39      M       engineer    55107

### Step 5. See the last 10 entries

In [ ]:
users.tail(10)

age gender     occupation zip_code
user_id                                    
934       61      M       engineer    22902
935       42      M         doctor    66221
936       24      M          other    32789
937       48      M       educator    98072
938       38      F     technician    55038
939       26      F        student    33319
940       32      M  administrator    02215
941       20      M        student    97229
942       48      F      librarian    78209
943       22      M        student    77841

### Step 6. What is the number of observations in the dataset?

In [ ]:
users.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 943 entries, 1 to 943
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   age         943 non-null    int64 
 1   gender      943 non-null    object
 2   occupation  943 non-null    object
 3   zip_code    943 non-null    object
dtypes: int64(1), object(3)
memory usage: 36.8+ KB


943 enteries

### Step 7. What is the number of columns in the dataset?

In [ ]:
users.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 943 entries, 1 to 943
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   age         943 non-null    int64 
 1   gender      943 non-null    object
 2   occupation  943 non-null    object
 3   zip_code    943 non-null    object
dtypes: int64(1), object(3)
memory usage: 36.8+ KB


4 columns

### Step 8. Print the name of all the columns.

In [ ]:
users.columns

Index(['age', 'gender', 'occupation', 'zip_code'], dtype='object')

### Step 9. How is the dataset indexed?

user_id: This is a unique identifier for each data record. It is the primary key for indexing individual records.

### Step 10. What is the data type of each column?

In [ ]:
print(f"type of age column is:", users["age"].dtype)
print(f"type of gender column is:", users["gender"].dtype)
print(f"type of occupation column is:", users["occupation"].dtype)
print(f"type of zip code column is:", users["zip_code"].dtype)

type of age column is: int64
type of gender column is: object
type of occupation column is: object
type of zip code column is: object


### Step 11. Print only the occupation column

In [ ]:
users["occupation"]

user_id
1         technician
2              other
3             writer
4         technician
5              other
           ...      
939          student
940    administrator
941          student
942        librarian
943          student
Name: occupation, Length: 943, dtype: object

### Step 12. How many different occupations are in this dataset?

In [ ]:
users["occupation"].nunique()

21

### Step 13. What is the most frequent occupation?

In [ ]:
users["occupation"].max()

'writer'

### Step 14. Summarize the DataFrame.

In [ ]:
users.describe()

age
count  943.000000
mean    34.051962
std     12.192740
min      7.000000
25%     25.000000
50%     31.000000
75%     43.000000
max     73.000000

### Step 15. Summarize all the columns

In [ ]:
users.columns

Index(['age', 'gender', 'occupation', 'zip_code'], dtype='object')

### Step 16. Summarize only the occupation column

In [ ]:
users["occupation"].describe()

count         943
unique         21
top       student
freq          196
Name: occupation, dtype: object

### Step 17. What is the mean age of users?

In [ ]:
users["age"].mean()

34.05196182396607

### Step 18. What is the age with least occurrence?

In [ ]:
age_counts = users['age'].value_counts()
min_occurrence_age = age_counts.idxmin()
min_occurrence_count = age_counts.min()

print(f"The age with the least occurrence is {min_occurrence_age} with {min_occurrence_count} occurrences.")

The age with the least occurrence is 7 with 1 occurrences.
